dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/consumer/TaskDispatchFailedQueueConsumer.java

```diff
@@ -0,0 +1,149 @@
+import org.apache.dolphinscheduler.server.master.config.MasterConfig;
+import org.apache.dolphinscheduler.server.master.metrics.TaskMetrics;
+import org.apache.dolphinscheduler.service.exceptions.TaskPriorityQueueException;
+import org.apache.dolphinscheduler.service.queue.TaskPriority;
+import org.apache.dolphinscheduler.service.queue.TaskPriorityQueue;
+import org.apache.dolphinscheduler.service.queue.TaskPriorityQueueImpl;
+
+import java.util.Objects;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.PostConstruct;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Component;
+
+@Component
+public class TaskDispatchFailedQueueConsumer extends BaseDaemonThread {
+
+    private static final Logger logger = LoggerFactory.getLogger(TaskDispatchFailedQueueConsumer.class);
+
+    /**
+     * taskPriorityQueue
+     */
+    @Autowired
+    @Qualifier(Constants.TASK_PRIORITY_QUEUE)
+    private TaskPriorityQueue<TaskPriority> taskPriorityQueueImpl;
+
+    /**
+     * taskDispatchFailedQueue
+     */
+    @Autowired
+    @Qualifier(Constants.TASK_DISPATCH_FAILED_QUEUE)
+    private TaskPriorityQueue<TaskPriority> taskDispatchFailedQueueImpl;
+
+    @Autowired
+    private MasterConfig masterConfig;
+
+    private ThreadPoolExecutor retryConsumerThreadPoolExecutor;
+
+    /**
+     * delay time for retries
+     */
+    private static final Long[] TIME_DELAY;
+
+    /**
+     * initialization failure retry delay rule
+     */
+    static {
+        TIME_DELAY = new Long[Constants.DEFAULT_MAX_RETRY_COUNT];
+        for (int i = 0; i < Constants.DEFAULT_MAX_RETRY_COUNT; i++) {
+            int delayTime = (i + 1) * 1000;
+            TIME_DELAY[i] = (long) delayTime;
+        }
+    }
+
+    protected TaskDispatchFailedQueueConsumer() {
+        super("TaskDispatchFailedQueueConsumerThread");
+    }
+
+    @PostConstruct
+    public void init() {
+        this.retryConsumerThreadPoolExecutor = (ThreadPoolExecutor) ThreadUtils
+                .newDaemonFixedThreadExecutor("TaskDispatchFailedQueueConsumerThread", masterConfig.getDispatchTaskNumber());
+        super.start();
+    }
+
+    @Override
+    public void run() {
+        while (!ServerLifeCycleManager.isStopped()) {
+            try {
+                failedRetry();
+            } catch (Exception e) {
+                TaskMetrics.incTaskDispatchError();
+                logger.error("failed task retry error", e);
+            }
+        }
+    }
+
+    public void failedRetry() throws TaskPriorityQueueException {
+        if (taskDispatchFailedQueueImpl.size() > 0) {
+            retryConsumerThreadPoolExecutor.submit(() -> dispatchFailedBackToTaskPriorityQueue(masterConfig.getDispatchTaskNumber()));
+        }
+    }
```

Comment:
This will cause OOM, if the `taskDispatchFailedQueueImpl` is not empty, it will submit a lot of task... until the task be trigger

You need to wait the before task finished, and then finish the current loop.

Reply:
Yes, thx, I will fix this.